### ISSUE TITLE
Suggestions on how to implement COM-centered flat-bottom restraint which applies to individual particles

### ISSUE THREAD

**Issue by stefdoerr** [openmm/openmm#4707]
So, I have the following problem. I want to restrain individual particles to be within X nanometers from the center of mass of a different molecule.

I have tried to come up with a solution with the given OpenMM forces but I cannot.  `CustomCentroidBondForce` is the closest I can find but it applies restaints between two COMs while I want to apply the restraint on the individual atoms of the first set.

I guess ideally I would want an `addParticle` option on `CustomCentroidBondForce` which adds individual particles and then access their `xyz` coordinates in the energy expression with `x`, `y`, `z` while the groups are numbered as they are already with `x1`, `x2` etc.

Is there any way to implement this with the existing forces though?

**Comment by stefdoerr** [comment]
The only other way I have come up with is to use a `CustomExternalForce` and define the COM as a global variable. But then I need to update the COM every few steps in the simulation which requires me to retrieve the coordinates from the GPU context which will slow down the simulation significantly if I do this frequently

**Comment by peastman** [comment]
In the CustomCentroidBondForce, make the individual particles each be its own group.  You'll have one group that's a whole molecule, and a lot of other groups for the individual atoms you want to restrain relative to its center of mass.

**Comment by peastman** [comment]
You write an energy expression that connects a single atom to the COM of the molecule.  Then you call `addBond()` once for each atom you want to restrain.

**Comment by stefdoerr** [comment]
Ok, I'll give it a try. 

Is there any way to get the periodic distance of a single axis though? I see that it has the `distance(g1, g2)` function but it doesn't seem to have the `periodicdistance(x1, y1, z1, x2, y2, z2)` function available so that I can use it like `periodicdistance(x1, 0, 0, x2, 0, 0)`.

**Comment by peastman** [comment]
Use the `pointdistance()` function.  If you call `setUsesPeriodicBoundaryConditions(True)`, it will be periodic.  Remember that's only valid for rectangular boxes.  For triclinic boxes, you can't separate out one axis from the others.

**Comment by stefdoerr** [comment]
Hi Peter, sorry to bother again, but could you provide us with an example for this? We are a bit stuck on how to write the energy function to "connect a single atom to the COM of the molecule". If you can provide a full example with how to add the bonds as well it would be super helpful. Thanks!

**Comment by peastman** [comment]
That looks right.  Think carefully about your energy function though.  What you've written has a singularity when x1=x2.

**Comment by stefdoerr** [comment]
There is something very weird going on with `CustomCentroidBondForce`. It seems to eat up all my memory.

```py
ff = mm.CustomCentroidBondForce(2, "k0 * ( (step(dist_z) * dist_z)^2 ); dist_z = pointdistance(0, 0, z2, 0, 0, z1) - wz00")
g0 = ff.addGroup(center_idx[:2000], masses[center_idx[:2000]])
for idx in sel_idx:
    gi = ff.addGroup([idx], [masses[idx]])
    ff.addBond([g0, gi])
ff.addGlobalParameter(
    f"k{i}", 10 * unit.kilocalories_per_mole / unit.angstrom**2,
)
ff.addGlobalParameter(f"wx0{i}", width[0] / 2 * unit.angstrom)
ff.addGlobalParameter(f"wy0{i}", width[1] / 2 * unit.angstrom)
ff.addGlobalParameter(f"wz0{i}", width[2] / 2 * unit.angstrom)
ff.setUsesPeriodicBoundaryConditions(True)
```

The more atoms are added to the first group the more memory is consumed. If I use 2000 atoms in the first group it uses 8GB of RAM, if I put 5000 atoms it crashes the python process due to running out of memory.

It's even more curious as the simulation should run on my CUDA GPU. So I don't know what exactly is stored in RAM.

It seems like a bug to me.

**Comment by stefdoerr** [comment]
I made a reproducible example here with a simple water box and a python OpenMM script which crashes out of memory on my machine. If you have more than 16GB it might not crash for you but just monitor the memory usage which spikes up a lot.

[reproducible_example.zip](https://github.com/user-attachments/files/17568285/reproducible_example.zip)

**Comment by peastman** [comment]
I see the problem.  It's in `ComputeContext::findMoleculeGroups()`, which works out the molecules based on what particles are bonded to each other.

https://github.com/openmm/openmm/blob/ba9113fb1d3736ab551f76ce3c3be31a2770078d/platforms/common/src/ComputeContext.cpp#L202-L222

Your force defines 1000 bonds, and each one involves 10,000 atoms, and for each of those it adds them to the vector for all 10,000 other atoms.  They really should be sets, not vectors, to avoid the duplication.

**Comment by peastman** [comment]
That fixes the memory use, but it's way too slow.  It's been running for almost 15 minutes and still going.  Let me see what else I can come up with.

**Comment by peastman** [comment]
A better solution is to change how the ForceInfo for CustomCentroidBondForce defines particle groups: one for every group, and one for each bond that contains only two particles.  I'll test it a bit more, then get in a PR with the change.

**Comment by stefdoerr** [comment]
Thanks for looking into it! 
I noticed that even with a single group it takes up lots of memory, like when trying to restrain the COM of my protein to a given position.
```py
    ff = mm.CustomCentroidBondForce(1, forcestr)
    g0 = ff.addGroup(sel_idx, masses[sel_idx])
    ff.addBond([g0])
```
This also brings my RAM usage from 6% up to 22% when reinitializing the context.

**Comment by peastman** [comment]
That's because of it making a list of which other atoms each atom is bonded to.  That code was written with the assumption that bonds only involve a few atoms, usually 2-4, not 10,000.

What about reducing the number of atoms in that group?  Instead of including all atoms, just include the heavy atoms, or just the alpha carbons, or something like that?  The center of mass of some well chosen subset should be nearly identical to the center of mass of all atoms.

**Comment by stefdoerr** [comment]
Yes it's possible, although it would be a shame as but it complicates the user work significantly. Now for example the user can just restrain a ligand to a specific distance to the COM of the whole protein with a trivial atom selection of "protein". In the other case they need to find some atoms which are close to the COM and also hope the COM doesn't change too much.

**Comment by stefdoerr** [comment]
> That's because of it making a list of which other atoms each atom is bonded to

But in the above case the atoms are not bonded to any other atoms, no? It's a single group

**Comment by stefdoerr** [comment]
Also it's a bit sneaky. If you are not aware of that limitation, your simulation setup crashes at a seemingly unrelated call far away from the actual problem, so it takes a bit to figure out where the problem is.

**Comment by peastman** [comment]
> But in the above case the atoms are not bonded to any other atoms, no? It's a single group

You still call `addBond()`.  That's a bond.  And every atom involved in the bond is considered to be bonded to every other, which in this case means every atom in the group whose position you're restraining.

**Comment by peastman** [comment]
I think I have a fix.  Can you test out #4713?

I think it should work as a good general purpose fix, but I need to make sure there are no cases where it leads to unexpected problems.

**Comment by stefdoerr** [comment]
Yes, seems like it worked! At least memory-wise I only saw a 0.1% increase after calling reinitialize on the context.

**Comment by peastman** [comment]
Fixed by #4713.

**Comment by dhruvakulkarni** [comment]
Hello! 

We are recently investigating moving from CHARMM to OpenMM for some coarse-grained MD runs, and we are very excited with some initial runs. However, we are running into an issue with a restraint force using CustomCentroidBondForce to calculate Rg to use for a RGYR Restraint Force (ala CHARMM).

We followed your suggestion of creating one group per atom, and one group with all atoms, and a bond for each single-atom group with the group containing all atoms within a CustomCentroidBondForce wrapped in CustomCVForce that does the harmonic restraint. This works when the number of atoms involved in the calc of the collective variable is in 100s.

When the number of atoms which are involved in the calculation of Rg  is beyond 1000, then the setup time becomes  slow (it just waits for ~30 mins without completing for example in the attached test case - `python scripts/md.py` should repr it). 

A profile for a successful run (around 1000 atoms) shows that most of this time, it is in the `findMoleculeGroups` function, and within that doing some inserts into a hashmap. (In the code it is at the `simulation = Simulation(modeller.topology, system, integrator)` step - hence I was digging around and thought this issue may be relevant?

We also have fixed and rigid body constraints in the system - so dont know if that is also contributing to this slowness. (I think I did try with setting mass to zero as well, but that made no difference)
There is an nsys profile, and a sample trace attached in the tar in case its helpful.

We would appreciate any suggestions and/or pointers. Thanks in advance!

[openmm-test.tar.gz](https://github.com/user-attachments/files/21203166/openmm-test.tar.gz)
